# 프로젝트 1-(1) : load_digits 손글씨를 분류해 봅시다.

In [1]:
# 필요한 모듈 import 하기

from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report


In [2]:
# 데이터 준비

digits = load_digits()
digits.keys()

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

In [3]:
# 데이터 이해하기

# Feature Data 지정하기

digits_data = digits.data
print(digits_data.shape)
digits_data[0]

(1797, 64)


array([ 0.,  0.,  5., 13.,  9.,  1.,  0.,  0.,  0.,  0., 13., 15., 10.,
       15.,  5.,  0.,  0.,  3., 15.,  2.,  0., 11.,  8.,  0.,  0.,  4.,
       12.,  0.,  0.,  8.,  8.,  0.,  0.,  5.,  8.,  0.,  0.,  9.,  8.,
        0.,  0.,  4., 11.,  0.,  1., 12.,  7.,  0.,  0.,  2., 14.,  5.,
       10., 12.,  0.,  0.,  0.,  0.,  6., 13., 10.,  0.,  0.,  0.])

In [4]:
# Label(Target) Data 지정하기

digits_target = digits.target
print(digits_target.shape)
digits_target [ : 30]

(1797,)


array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
       2, 3, 4, 5, 6, 7, 8, 9])

In [5]:
# Target Names 출력해 보기

digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [6]:
# 데이터 Describe 해보기

print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 5620
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

In [7]:
# train, test 데이터 분리

X_train, X_test, y_train, y_test = train_test_split(digits_data, 
                                                    digits_target,
                                                    test_size = 0.2,
                                                    random_state = 16)

### <다양한 모델로 학습시켜보기>

In [8]:
# DecisionTree 사용해보기

from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state = 16)
decision_tree.fit(X_train, y_train)
decision_tree_y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, decision_tree_y_pred))

              precision    recall  f1-score   support

           0       0.94      0.89      0.91        35
           1       0.72      0.76      0.74        34
           2       0.83      0.86      0.85        35
           3       0.79      0.81      0.80        37
           4       0.84      0.79      0.82        39
           5       0.86      0.80      0.83        30
           6       0.81      0.97      0.88        35
           7       0.95      0.80      0.86        44
           8       0.79      0.72      0.75        36
           9       0.72      0.83      0.77        35

    accuracy                           0.82       360
   macro avg       0.82      0.82      0.82       360
weighted avg       0.83      0.82      0.82       360



In [9]:
# RandomForest 사용해 보기

from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier()
random_forest.fit(X_train, y_train)
random_forest_y_pred = random_forest.predict(X_test)

print(classification_report(y_test, random_forest_y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        35
           1       0.97      1.00      0.99        34
           2       0.97      1.00      0.99        35
           3       0.97      0.97      0.97        37
           4       0.95      1.00      0.97        39
           5       1.00      0.97      0.98        30
           6       1.00      1.00      1.00        35
           7       0.98      0.98      0.98        44
           8       0.97      0.92      0.94        36
           9       0.97      0.94      0.96        35

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



In [10]:
# SVM 사용해 보기

from sklearn import svm

svm_model = svm.SVC()
print(svm_model._estimator_type)

svm_model.fit(X_train, y_train)
svm_y_pred = svm_model.predict(X_test)

print(classification_report(y_test, svm_y_pred))

classifier
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        35
           1       0.97      1.00      0.99        34
           2       1.00      1.00      1.00        35
           3       1.00      1.00      1.00        37
           4       1.00      1.00      1.00        39
           5       1.00      1.00      1.00        30
           6       1.00      1.00      1.00        35
           7       1.00      0.98      0.99        44
           8       0.97      0.97      0.97        36
           9       0.97      0.97      0.97        35

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



In [11]:
# SGD Classifier 사용해 보기

from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
sgd_y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, sgd_y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        35
           1       0.79      0.91      0.85        34
           2       1.00      0.94      0.97        35
           3       0.88      1.00      0.94        37
           4       0.97      0.97      0.97        39
           5       1.00      0.90      0.95        30
           6       1.00      1.00      1.00        35
           7       0.95      0.93      0.94        44
           8       1.00      0.83      0.91        36
           9       0.89      0.94      0.92        35

    accuracy                           0.94       360
   macro avg       0.95      0.94      0.94       360
weighted avg       0.95      0.94      0.95       360



In [12]:
# Logistic Regression 사용해 보기

from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression() 
logistic_model.fit(X_train, y_train)
logistic_model_y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, logistic_model_y_pred))

              precision    recall  f1-score   support

           0       1.00      0.94      0.97        35
           1       0.87      0.97      0.92        34
           2       0.97      0.94      0.96        35
           3       1.00      0.97      0.99        37
           4       0.95      0.95      0.95        39
           5       0.88      0.97      0.92        30
           6       1.00      1.00      1.00        35
           7       0.98      0.93      0.95        44
           8       0.91      0.89      0.90        36
           9       0.97      0.97      0.97        35

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### <모델 평가하기>

1. load_digits() 데이터는 손글씨를 얼만큼 정확하게 인식은 잘했는지가 중요한 관건이다.
  - 손글씨로 쓴 숫자를 정확한 숫자로 인식한 결과값만이 의미가 있기 때문에 recall(재현율) = TP / TP + FN의 값을 비교하여 좋은 모델을 선택할 것이다.      
     
     
2. 아래와 같이 sklearn.metrics에서 제공하는 recall_score 평가지표를 사용했다.

In [14]:
from sklearn.metrics import recall_score

print('Decision Tree : {}'.format(recall_score(y_test, decision_tree_y_pred, average = 'weighted')))
print('Random Forest : {}'.format(recall_score(y_test, random_forest_y_pred, average = 'weighted')))
print('SVM : {}'.format(recall_score(y_test, svm_y_pred, average = 'weighted')))
print('SGD : {}'.format(recall_score(y_test, sgd_y_pred, average = 'weighted')))
print('Logistic Regression : {}'.format(recall_score(y_test, logistic_model_y_pred, average = 'weighted')))

Decision Tree : 0.8222222222222222
Random Forest : 0.9777777777777777
SVM : 0.9916666666666667
SGD : 0.9444444444444444
Logistic Regression : 0.9527777777777777


### <load_digits 데이터의 recall_score 결과>

1. 위의 recall_score의 결과에서 볼 수 있듯이 SVM 모델이 0.991로 가장 score가 높기 때문에 load_digits의 데이터 예측을 가장 잘하는 모델로 선택할 수 있다.   

# 프로젝트 1-(2) load_wine 와인을 분류해 봅시다.

In [15]:
# 필요한 모듈 import 하기

from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report


In [16]:
# 데이터 준비

wine = load_wine()
type(wine)

sklearn.utils.Bunch

In [17]:
# 데이터 이해하기

# Feature Data 지정하기

wine_data = wine.data
print(wine_data.shape)
print(wine_data.ndim)   # .ndim은 feature data가 몇 차원으로 이루어져 있는지 알 수 있음
wine_data[0]

(178, 13)
2


array([1.423e+01, 1.710e+00, 2.430e+00, 1.560e+01, 1.270e+02, 2.800e+00,
       3.060e+00, 2.800e-01, 2.290e+00, 5.640e+00, 1.040e+00, 3.920e+00,
       1.065e+03])

In [18]:
# Label Data 지정하기

wine_target = wine.target
print(wine_target.shape)
wine_target[:10]

(178,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [19]:
wine.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

In [20]:
# Target Names 출력해 보기

print(wine.target_names)

['class_0' 'class_1' 'class_2']


In [21]:
# 데이터 Describe 해보기

print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

In [22]:
# train, test 데이터 분리하기

X_train, X_test, y_train, y_test = train_test_split(wine_data,
                                                    wine_target,
                                                    test_size = 0.2,
                                                    random_state = 321)

### < 다양한 모델로 학습시켜보기 >

In [28]:
# Decision Tree 사용해 보기

from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier()
decision_tree.fit(X_train, y_train)
decision_tree_y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, decision_tree_y_pred))

              precision    recall  f1-score   support

           0       0.82      0.82      0.82        11
           1       0.89      0.89      0.89        19
           2       1.00      1.00      1.00         6

    accuracy                           0.89        36
   macro avg       0.90      0.90      0.90        36
weighted avg       0.89      0.89      0.89        36



In [29]:
# Random Forest 사용해 보기

from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier()
random_forest.fit(X_train, y_train)
random_forest_y_pred = random_forest.predict(X_test)

print(classification_report(y_test, random_forest_y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        11
           1       1.00      1.00      1.00        19
           2       1.00      1.00      1.00         6

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



In [30]:
# SVM 사용해 보기

from sklearn import svm

svm_model = svm.SVC()
print(svm_model._estimator_type)
svm_model.fit(X_train, y_train)
svm_y_pred = svm_model.predict(X_test)

print(classification_report(y_test, svm_y_pred))

classifier
              precision    recall  f1-score   support

           0       0.90      0.82      0.86        11
           1       0.86      0.63      0.73        19
           2       0.33      0.67      0.44         6

    accuracy                           0.69        36
   macro avg       0.70      0.71      0.68        36
weighted avg       0.78      0.69      0.72        36



In [31]:
# SGD Classifier 사용해 보기

from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
print(sgd_model._estimator_type)
sgd_model.fit(X_train, y_train)
sgd_y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, sgd_y_pred))

classifier
              precision    recall  f1-score   support

           0       0.59      0.91      0.71        11
           1       0.74      0.74      0.74        19
           2       0.00      0.00      0.00         6

    accuracy                           0.67        36
   macro avg       0.44      0.55      0.48        36
weighted avg       0.57      0.67      0.61        36



/opt/conda/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [32]:
# Logistic Regression 사용해 보기

from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression(max_iter=5000)  # 반복횟수의 max를 크게하지 않으면 경고 창이 뜨기 때문에 구글링을 하여 max_iter 기본값인 100보다 크게 설정하였음.
logistic_model.fit(X_train, y_train)
logistic_model_y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, logistic_model_y_pred))

              precision    recall  f1-score   support

           0       1.00      0.91      0.95        11
           1       0.95      1.00      0.97        19
           2       1.00      1.00      1.00         6

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.98        36
weighted avg       0.97      0.97      0.97        36



### <모델 평가하기>

1. load_wine 데이터도 load_digits데이터와 마찬가지로 와인 특성에 따라 정확하게 와인 종류로 분류한 값만이 의미가 있다고 생각하기 때문에 recall(재현율) score로 결과를 알아볼 것이다.    

In [33]:
# sklear.metrics의 recall_score로 결과 비교하기

from sklearn.metrics import recall_score

print('Decision Tree : {}'.format(recall_score(y_test, decision_tree_y_pred, average = 'weighted')))
print('Random Forest : {}'.format(recall_score(y_test, random_forest_y_pred, average = 'weighted')))
print('SVM : {}'.format(recall_score(y_test, svm_y_pred, average = 'weighted')))
print('SGD : {}'.format(recall_score(y_test, sgd_y_pred, average = 'weighted')))
print('Logistic Regression : {}'.format(recall_score(y_test, logistic_model_y_pred, average = 'weighted')))

Decision Tree : 0.8888888888888888
Random Forest : 1.0
SVM : 0.6944444444444444
SGD : 0.6666666666666666
Logistic Regression : 0.9722222222222222


### < load_wine데이터의 recall_score 결과>

1. load_wine 데이터는 여러가지 와인 특성에 따라 종류로 분류되기 때문에 의사결정트리인 Decision Tree나 의사결정트리를 여러 개 모아놓은 Random Forest가 좋은 모델이 될 수 있을거라고 예상했는데 recall_score를 보니 역시나 Random Forest가 1.0이라는 결과가 나왔다. 

#  프로젝트 1-(3) load_breast_cancer 유방암 여부를 진단해 봅시다.

In [34]:
# 필요한 모듈 import 하기

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report


In [35]:
# 데이터 준비

cancer = load_breast_cancer()
type(cancer)
cancer.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [36]:
# 데이터 이해하기

# Feature Data 지정하기

cancer_data = cancer.data
print(cancer_data.shape)
cancer_data[0]

(569, 30)


array([1.799e+01, 1.038e+01, 1.228e+02, 1.001e+03, 1.184e-01, 2.776e-01,
       3.001e-01, 1.471e-01, 2.419e-01, 7.871e-02, 1.095e+00, 9.053e-01,
       8.589e+00, 1.534e+02, 6.399e-03, 4.904e-02, 5.373e-02, 1.587e-02,
       3.003e-02, 6.193e-03, 2.538e+01, 1.733e+01, 1.846e+02, 2.019e+03,
       1.622e-01, 6.656e-01, 7.119e-01, 2.654e-01, 4.601e-01, 1.189e-01])

In [37]:
# Label(=Target) Data 지정하기

cancer_target = cancer.target
print(cancer_target.shape)
cancer_target[ : 10]

(569,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [38]:
# Target Names 출력해 보기

print(cancer.target_names)

['malignant' 'benign']


In [39]:
# 데이터 Desceribe 해 보기

print(cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

In [47]:
# train, test 데이터 분리

X_train, X_test, y_train, y_test = train_test_split(cancer_data,
                                                    cancer_target,
                                                    test_size = 0.2,
                                                    random_state = 321)


### < 다양한 모델로 학습시켜보기>

In [48]:
# Decision Tree 사용해보기

from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier()
decision_tree.fit(X_train, y_train)
decision_tree_y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, decision_tree_y_pred))

              precision    recall  f1-score   support

           0       0.87      0.96      0.91        49
           1       0.97      0.89      0.93        65

    accuracy                           0.92       114
   macro avg       0.92      0.93      0.92       114
weighted avg       0.93      0.92      0.92       114



In [49]:
# Random Forest 사용해 보기

from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier()
random_forest.fit(X_train, y_train)
random_forest_y_pred = random_forest.predict(X_test)

print(classification_report(y_test, random_forest_y_pred))

              precision    recall  f1-score   support

           0       0.96      0.94      0.95        49
           1       0.95      0.97      0.96        65

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.96       114
weighted avg       0.96      0.96      0.96       114



In [50]:
# SVM 사용해 보기

from sklearn import svm

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
svm_y_pred = svm_model.predict(X_test)

print(classification_report(y_test, svm_y_pred))

              precision    recall  f1-score   support

           0       0.95      0.80      0.87        49
           1       0.86      0.97      0.91        65

    accuracy                           0.89       114
   macro avg       0.91      0.88      0.89       114
weighted avg       0.90      0.89      0.89       114



In [51]:
# SGD Classifier 사용해 보기

from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
sgd_y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, sgd_y_pred))

              precision    recall  f1-score   support

           0       1.00      0.84      0.91        49
           1       0.89      1.00      0.94        65

    accuracy                           0.93       114
   macro avg       0.95      0.92      0.93       114
weighted avg       0.94      0.93      0.93       114



In [53]:
# Logistic Regression 사용해 보기

from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression()  
logistic_model.fit(X_train, y_train)
logistic_model_y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, logistic_model_y_pred))

              precision    recall  f1-score   support

           0       0.93      0.82      0.87        49
           1       0.87      0.95      0.91        65

    accuracy                           0.89       114
   macro avg       0.90      0.89      0.89       114
weighted avg       0.90      0.89      0.89       114



/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### < 모델 평가하기 >

1. 암을 진단하는 경우 실제 환자를 한 명이라도 놓치면 안 좋기 때문에 정확한 판단을 위해 recall_score가 중요하다.

In [56]:
from sklearn.metrics import recall_score

print('Decision Tree : {}'.format(recall_score(y_test, decision_tree_y_pred, average = 'weighted')))
print('Random Forest : {}'.format(recall_score(y_test, random_forest_y_pred, average = 'weighted')))
print('SVM : {}'.format(recall_score(y_test, svm_y_pred, average = 'weighted')))
print('SGD : {}'.format(recall_score(y_test, sgd_y_pred, average = 'weighted')))
print('Logistic Regression : {}'.format(recall_score(y_test, logistic_model_y_pred, average = 'weighted')))

Decision Tree : 0.9210526315789473
Random Forest : 0.956140350877193
SVM : 0.8947368421052632
SGD : 0.9298245614035088
Logistic Regression : 0.8947368421052632


### < load_breast_cancer 데이터의 recall_score 결과>

1. 암 진단은 양성/음성 이진 분류 문제라고 생각했기에 당연히 Logistic Regression의 결과가 가장 좋게 나올 거라고 예상했지만 나의 예상과 달리 Random Forest모델의 결과가 0.956으로 가장 높게 나왔다.       
         
     
2. 위의 와인도 feature의 종류가 많아서 Decision Tree와 Random Forest의 결과가 높게 나온 것처럼 유방암의 특성 갯수가 많아서 Random Forest의 성능이 더 높게 나온 것 같다.     


# 회고

### <느낀점>


1. 먼저 scikit-learn에서 제공된 데이터로 분류를 하게 되어 지난 번 Ex1) 가위바위보 프로젝트보다 데이터에 신경을 쓰지 않아서 좋았다.   
      
           
2. train data와 test data를 따로 만들지 않고 sklearn의 train_test_split함수를 이용해 한 번에 쉽게 나눌 수 있어서 편했다.    
        
             
3. 머신러닝의 다양한 모델 학습 방법들이 있는지 처음 알았다. 또한, 학습된 모델들의 test data 예측 결과를 여러가지 방면으로 해석해야 하는 것도 처음 알게 되었다.      
 - 정확도만 좋으면 성능이 좋은 줄 알았지만 오차행렬에 대한 개념도 알아야 하고 데이터의 특징에 따라 precision, recall, F1-score 등 다양하게 해석해야 하는 걸 이번에 알게 되었다.      
       

# 결론

### * 파이썬에서 많이 사용하고 있는 머신러닝 라이브러리인 scikit-learn을 통해 데이터 학습에 여러 가지 모델을 적용할 수 있고 쉽고 빠르게 원하는 결과를 얻을 수 있다.